<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
### YOUR CODE STARTS HERE
from scipy.stats import ttest_ind
import pandas as pd

In [2]:
# Load data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-01-21 02:26:12--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2020-01-21 02:26:17 (639 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [5]:
# Create the dataframe
col_headers = ['party', 'handicapped-infants', 'water-project', 'budget', 
               'physician-fee-freeze', 'el-salvador-aid', 'religious-groups', 
               'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 
               'immigration', 'synfuels', 'education', 'right-to-sue', 'crime', 
               'duty-free', 'south-africa']
df = pd.read_csv('house-votes-84.data', header=None, 
                 names=col_headers, na_values='?')
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
# replace y and n with 1 and 0
df = df.replace({'y': 1, 'n': 0})

In [0]:
# split dem and rep votes
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']

In [0]:
# function to easily t-test with
def run_t_test(vote_name):
  print("Dem votes for: " + str(dem[vote_name].sum()))
  print("Rep votes for: " + str(rep[vote_name].sum()))
  print(ttest_ind(dem[vote_name], rep[vote_name], nan_policy='omit'))


In [27]:
run_t_test('budget')

Dem votes for: 231.0
Rep votes for: 22.0
Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)


## Support for the budget bill

Due to a p-value of 2.07e-77, we can conclude with over 99% certainty that the two parties had different levels of support for the budget bill.

In [28]:
run_t_test('religious-groups')

Dem votes for: 123.0
Rep votes for: 149.0
Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)


## Support for the religous groups bill

Due to a p-value of 2.39e-20, we can conclude with over 99% certainty that the two parties had different levels of support for the religous groups bill.

In [43]:
run_t_test('water-project')

Dem votes for: 120.0
Rep votes for: 75.0
Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)


## Support for the water project bill

Due to a p-value of .929, we cannot conclude that the two parties had differing levels of support for the water project bill.

In [42]:
# Because I wanted to see them all
for col in rep.columns:
  if(col != 'party'):
    print(col)
    run_t_test(col)

handicapped-infants
Dem votes for: 156.0
Rep votes for: 31.0
Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)
water-project
Dem votes for: 120.0
Rep votes for: 75.0
Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)
budget
Dem votes for: 231.0
Rep votes for: 22.0
Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)
physician-fee-freeze
Dem votes for: 14.0
Rep votes for: 163.0
Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)
el-salvador-aid
Dem votes for: 55.0
Rep votes for: 157.0
Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)
religious-groups
Dem votes for: 123.0
Rep votes for: 149.0
Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)
anti-satellite-ban
Dem votes for: 200.0
Rep votes for: 39.0
Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)
aid-to-contras
Dem votes for: 218.0
Rep votes for: 24.0
Ttest_indR